[1.12] advancements pagination

Xalcon opened the pull request:
By default, minecraft only supports 26 advancement tabs (8 on top and bottom, 5 on left and right).
This PR adds 2 buttons to the advancements gui at the top of the screen to allow an "unlimited" amount of advancement tabs.

It also includes a test mod that adds 64 dummy advancement tabs to allow testing of the feature.

I tried to do this similar to the creative tab, so this might not be the cleanest way to do this, andI'm not really happy with the current solution. I didnt want to do more patches to the vanilla code than needed, so I just submit this here.

![example](http://i.imgur.com/HHvZ3jY.png)

Xalcon committed:
add pagination to GuiScreenAdvancements

Xalcon committed:
add debug mod with 64 dummy advancements

ichttt commented on patches/minecraft/net/minecraft/client/gui/advancements/GuiScreenAdvancements.java.patch:
```diff
@@ -0,0 +1,100 @@
+ import javax.annotation.Nullable;
+ import net.minecraft.advancements.Advancement;
+ import net.minecraft.advancements.AdvancementProgress;
++import net.minecraft.client.gui.GuiButton;
```
no imports in patches

ichttt commented on patches/minecraft/net/minecraft/client/gui/advancements/GuiScreenAdvancements.java.patch:
```diff
@@ -0,0 +1,100 @@
+     private int field_191942_u;
+     private boolean field_191944_v;
++    private static int tabPage;
++    private static int maxPages;
```
this can be one line: private static int tabPage, maxPages;

ichttt commented on patches/minecraft/net/minecraft/client/gui/advancements/GuiScreenAdvancements.java.patch:
```diff
@@ -0,0 +1,100 @@
+         return this.field_191947_i.get(p_191935_1_);
+     }
++
++    protected void func_146284_a(GuiButton button) throws IOException
```
missing override

ichttt commented on patches/minecraft/net/minecraft/client/gui/advancements/GuiScreenAdvancements.java.patch:
```diff
@@ -0,0 +1,100 @@
+             {
+-                guiadvancementtab1.func_191796_a(p_191934_1_, p_191934_2_, this.field_146296_j);
++                if(guiadvancementtab1.getPage() == tabPage)
++                    guiadvancementtab1.func_191796_a(p_191934_1_, p_191934_2_, this.field_146296_j);
```
this should be one line

marvin-roesch replied:
If you don't change the indentation, the patch will only be a one line addition instead of a deletion and addition.

Xalcon replied:
ah, saw this too late, will do.

Xalcon committed:
patch cleanup (remove static import, move if body to same line, etc)

Xalcon committed:
move if-body back to its own line without indenting it

"If you don't change the indentation, the patch will only be a one line addition instead of a deletion and addition." - *PaleoCrafter*

bs2609 commented:
Does this need a mod to test it, can't this be tested using the vanilla method for custom advancements?

Xalcon commented:
Yes, you could move the advancements to the worlds advancement directory, but I thought its more convenient as a mod.

mezz commented on patches/minecraft/net/minecraft/client/gui/advancements/GuiAdvancementTab.java.patch:
```diff
@@ -0,0 +1,47 @@
+     private boolean field_192992_s;
++    private int page;
+ 
++    public GuiAdvancementTab(Minecraft p_i47589_1_, GuiScreenAdvancements p_i47589_2_, AdvancementTabType p_i47589_3_, int p_i47589_4_, int page, Advancement p_i47589_5_, DisplayInfo p_i47589_6_)
```
Please add new methods/constructors to the end of the file in a `/***** FORGE *****/` section like some other patches.

Xalcon replied:
done

mezz commented on patches/minecraft/net/minecraft/client/gui/advancements/GuiAdvancementTab.java.patch:
```diff
@@ -0,0 +1,47 @@
+         else
+         {
++            int page = p_193936_2_ / AdvancementTabType.field_192659_e;
++            int indexOnPage = p_193936_2_ % AdvancementTabType.field_192659_e;
```
I think you could reuse the `p_193936_2_` variable here instead of introducing a new one (`indexOnPage`), doing that will eliminate a couple patch lines below.

Xalcon replied:
i removed `page` and `indexOnPage`; they are now calculated on the fly :)

mezz reviewed (changes requested):
Looks good, might be able to get the patch smaller.

Xalcon committed:
reduce amount of patches done to the GuiAdvancementTab class.

as requested by mezz

Xalcon committed:
add *forge end* to the forge section

mezz commented:
Tested and looks good to me.
